chain-impl-mockchain/src/ledger.rs

```diff
@@ -865,77 +895,299 @@ pub mod test {
+                address: user2_address.clone(),
+                value: Value(1),
+            })
+            .finalize()
+            .with_account_witness(&block0_hash, &SpendingCounter::zero(), &sk1)
+            .seal();
+
+        let dyn_params = ledger.get_ledger_parameters();
+        let r = ledger.apply_transaction(&signed_tx, &dyn_params);
+        assert!(r.is_ok());
+    }
+
+    #[test]
+    pub fn account_to_delegation_correct_transaction() {
+        let discrimination = Discrimination::Test;
+
+        let mut rng = rand::thread_rng();
+        let mut delegation_rng = rand::thread_rng();
+        let (sk1, pk1, user1_address) = make_account_key(&mut rng, &discrimination);
+        let (_sk2, _pk2, user2_address) =
+            make_utxo_delegation_key(&mut rng, &mut delegation_rng, &discrimination);
+
+        let (message, _) = create_initial_transaction(Output {
+            address: user1_address.clone(),
+            value: Value(42000),
+        });
+
+        let (block0_hash, ledger) = create_initial_fake_ledger(discrimination, &[message]);
+        let signed_tx = TransactionBuilder::new()
+            .with_input(Input::from_account(
+                AccountIdentifier::from_single_account(account::Identifier::from(pk1)),
+                Value(1),
+            ))
+            .with_output(Output {
+                address: user2_address.clone(),
+                value: Value(1),
+            })
+            .finalize()
+            .with_account_witness(&block0_hash, &SpendingCounter::zero(), &sk1)
+            .seal();
+
+        let dyn_params = ledger.get_ledger_parameters();
+        let r = ledger.apply_transaction(&signed_tx, &dyn_params);
+        assert!(r.is_ok());
+    }
+
+    #[test]
+    pub fn delegation_to_account_correct_transaction() {
+        let discrimination = Discrimination::Test;
+
+        let mut rng = rand::thread_rng();
+        let mut delegation_rng = rand::thread_rng();
+        let (sk1, _pk1, user1_address) =
+            make_utxo_delegation_key(&mut rng, &mut delegation_rng, &discrimination);
+        let (_sk2, _pk2, user2_address) = make_account_key(&mut rng, &discrimination);
+
+        let (message, utxos) = create_initial_transaction(Output {
+            address: user1_address.clone(),
+            value: Value(42000),
+        });
+
+        let (block0_hash, ledger) = create_initial_fake_ledger(discrimination, &[message]);
+
+        let signed_tx = TransactionBuilder::new()
+            .with_input(Input::from_utxo(utxos[0]))
+            .with_output(Output {
+                address: user2_address.clone(),
+                value: Value(1),
+            })
+            .finalize()
+            .with_utxo_witness(&block0_hash, &sk1)
+            .seal();
+
+        let dyn_params = ledger.get_ledger_parameters();
+        let r = ledger.apply_transaction(&signed_tx, &dyn_params);
+        assert!(r.is_ok());
+    }
+    struct TransactionBuilder {
```

Comment:
This is great, but would it make sense to either unify with an existing functionality or move this to a general module ?